pkg/node-labels-resources/grpc_server.go

```diff
@@ -0,0 +1,341 @@
+		return
+	}
+	podNodeName := podDeleted.Spec.NodeName
+	nodeDetail, ok := nlm.resourceLists[podNodeName]
+	if ok {
+		delete(nodeDetail.Pods, podDeleted.Name)
+		nlm.resourceLists[podNodeName] = nodeDetail
+	} else {
+		klog.V(4).Infof("OnPodDelete: Failed to find node %s in resourceLists", podNodeName)
+	}
+	nlm.writeClusterResources()
+}
+
+func (nlm *NodeLabelsMonitor) writeClusterResources() {
+	podResourceUsage := corev1.ResourceList{}
+	nodeAllocatable := corev1.ResourceList{}
+	for _, nodeDetail := range nlm.resourceLists {
+		if !nodeDetail.Schedulable {
+			continue
+		}
+		addResources(nodeAllocatable, nodeDetail.Allocatable)
+		for _, podResource := range nodeDetail.Pods {
+			addResources(podResourceUsage, podResource)
+		}
+	}
+	subResources(nodeAllocatable, podResourceUsage)
+	nlm.nodeMutex.Lock()
+	nlm.allocatable = nodeAllocatable.DeepCopy()
+	klog.V(4).Infof("Cluster resources: %v", nlm.allocatable)
+	nlm.nodeMutex.Unlock()
+	err := nlm.NotifyChange(nlm.ctx, &resourcemonitors.ClusterIdentity{ClusterID: resourcemonitors.AllClusterIDs})
+	if err != nil {
+		log.Printf("error: error during sending notification to liqo: %s", err)
+	}
+}
+
+func extractPodResources(podToExtract *corev1.Pod) corev1.ResourceList {
+	resourcesToExtract, _ := resourcehelper.PodRequestsAndLimits(podToExtract)
+	return resourcesToExtract
+}
+
+func (nlm *NodeLabelsMonitor) ReadResources(ctx context.Context, req *resourcemonitors.ClusterIdentity) (*resourcemonitors.PoolResourceList, error) {
+	klog.V(4).Infof("info: reading resources for cluster %s", req.ClusterID)
+	nlm.nodeMutex.RLock()
+	defer nlm.nodeMutex.RUnlock()
+	allocatable := make(map[string]*resource.Quantity)
+	for resName, quantity := range nlm.allocatable {
+		val := quantity.DeepCopy()
+		allocatable[resName.String()] = &val
+	}
+	klog.V(4).Infof("Cluster resources: %v", allocatable)
+
+	resourceList := []*resourcemonitors.ResourceList{{Resources: allocatable}}
+	return &resourcemonitors.PoolResourceList{ResourceLists: resourceList}, nil
+}
+
+func (nlm *NodeLabelsMonitor) Subscribe(req *resourcemonitors.Empty, srv resourcemonitors.ResourceReader_SubscribeServer) error {
+	klog.V(1).Infof("info: liqo controller manager subscribed")
+
+	// Store the stream. Using req as key since each request will have a different req object.
+	nlm.subscribers.Store(req, srv)
+	ctx := srv.Context()
+
+	// This notification is useful since you can edit the resources declared in the deployment and apply it to the cluster when one or more
+	// foreign clusters are already peered so this broadcast notification will update the resources for those clusters.
+	err := nlm.NotifyChange(ctx, &resourcemonitors.ClusterIdentity{ClusterID: resourcemonitors.AllClusterIDs})
+	if err != nil {
+		klog.V(1).Infof("error: error during sending notification to liqo: %s", err)
+	}
+
+	for {
+		<-ctx.Done()
+		nlm.subscribers.Delete(req)
+		klog.V(1).Infof("info: liqo controller manager disconnected")
+		return nil
+	}
```

Comment:
```suggestion
    <-ctx.Done()
    nlm.subscribers.Delete(req)
    klog.V(1).Infof("info: liqo controller manager disconnected")
    return nil
```

The for loop is not required here